### PR TITLE
Add S3 exporter

### LINKED ISSUE #1407: Add metrics for etcd backups (S3)
### Current state
For each user-cluster a CronJob exists, which is responsible for creating a snapshot of the user-cluster etcd & afterwards to store it.
The whole storing of the snapshot is highly configurable.
Proposal: https://github.com/kubermatic/kubermatic/blob/master/docs/proposals/etcd-backup.md

By default we offer a chart for Minio & provide a Store-Container which supports storing to S3.

### Problem
We have no monitoring in place for the backups.

### Desired state
We have monitoring in place for the backups.
Ideally we have some kind of S3 exporter as this is the default solution.
Using a minio exporter might be too limited, as customers probably wan't to use a custom out-of-cluster S3 solution.

### Scope
This issue just focuses on getting the exporter in place.
Alerts & Dashboards should be done in separate issues

### PR DESCRIPTION
**Pull request by alvaroaleman**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1407 

**Special notes for your reviewer**:

**Release note**:
```release-note monitoring
Added an S3 exporter for metrics
```

### CODE REVIEW

**Review by kron4eg — CHANGES_REQUESTED** [review]
is that `api/cmd/s3-exporter/s3-exporter` a compiled binary?

**Comment by alvaroaleman** [comment]
@kron4eg Yes, `api/cmd/s3-exporter/s3-exporter` is a compiled binary after executing `api/hack/publish-s3-exporter.sh`

**Comment by kron4eg** [comment]
@alvaroaleman I mean, let's not commit to git compiled executables

**Comment by alvaroaleman** [comment]
@kron4eg Oups. Fixed

File: api/pkg/exporters/s3/exporter.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	ObjectsCount            *prometheus.GaugeVec
+	ObjectsLastModifiedDate *prometheus.GaugeVec
+	ObjectsEmptyCount       *prometheus.GaugeVec
+	QuerySuccess            prometheus.Gauge
+	bucket                  string
+	minioClient             *minio.Client
+}
+
+// MustRun starts a s3 exporter or panic
+func MustRun(minioClient *minio.Client, bucket string, listenAddress int) error {
```
Comment (mrIncompetent) [inline api/pkg/exporters/s3/exporter.go]: So why return also an error?

File: api/cmd/s3-exporter/main.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+	secure := true
+	if strings.HasPrefix(*endpointWithProto, "http://") {
+		glog.Info("Disabling tls due to http:// prefix in url..")
+		secure = false
+	}
+	endpoint := strings.TrimPrefix(*endpointWithProto, "http://")
+	endpoint = strings.TrimPrefix(endpoint, "https://")
+
+	stopChannel := make(chan struct{})
```
Comment (mrIncompetent) [inline api/cmd/s3-exporter/main.go:51]: what is this channel for?

File: api/cmd/s3-exporter/main.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/golang/glog"
+	"github.com/minio/minio-go"
+)
+
+func main() {
+	endpointWithProto := flag.String("endpoint", "", "https://my-s3.com:9000")
+	accessKeyID := flag.String("access-key-id", "", "S3 Access key")
+	secretAccessKey := flag.String("secret-access-key", "", "S3 Secret Access Key")
+	bucket := flag.String("bucket", "kubermatic-etcd-backups", "The bucket to monitor")
+	listenPort := flag.Int("listen-port", 9340, "The port to listen on")
```
Comment (mrIncompetent) [inline api/cmd/s3-exporter/main.go]: Please adjust to the existing way we define flags:
https://github.com/kubermatic/kubermatic/blob/master/api/cmd/kubermatic-api/main.go#L74

File: api/cmd/s3-exporter/main.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	endpoint = strings.TrimPrefix(endpoint, "https://")
+
+	stopChannel := make(chan struct{})
+	minioClient, err := minio.New(endpoint, *accessKeyID, *secretAccessKey, secure)
+	if err != nil {
+		glog.Fatalf("Failed to get S3 client: %v", err)
+	}
+
+	if err = s3.MustRun(minioClient, *bucket, *listenPort); err != nil {
+		glog.Fatalf(err.Error())
```
Comment (mrIncompetent) [inline api/cmd/s3-exporter/main.go]:
```
glog.Fatal(err)
```
should be enough?

File: api/cmd/s3-exporter/README.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    	log to standard error instead of files
+  -secret-access-key string
+    	S3 Secret Access Key
+  -stderrthreshold value
+    	logs at or above this threshold go to stderr
+  -v value
+    	log level for V logs
+  -vmodule value
+    	comma-separated list of pattern=N settings for file-filtered logging
+```
```
Comment (mrIncompetent) [inline api/cmd/s3-exporter/README.md:37]: instruction on building & releasing is missing

File: api/hack/publish-s3-exporter.sh [hunk 6]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+export REGISTRY=docker.io/kubermatic/s3-exporter
```
Comment (mrIncompetent) [inline api/hack/publish-s3-exporter.sh]: We are transitioning to quay.io. We should put new images only on quay.

I created docker quay.io/kubermatic/s3-exporter

File: api/pkg/exporters/s3/exporter.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+	if err := registry.Register(exporter.ObjectsEmptyCount); err != nil {
+		return err
+	}
+	if err := registry.Register(exporter.QuerySuccess); err != nil {
+		return err
+	}
+
+	promHandler := promhttp.HandlerFor(registry, promhttp.HandlerOpts{})
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		if !exporter.refreshMetrics(w, r) {
```
Comment (mrIncompetent) [inline api/pkg/exporters/s3/exporter.go]: I like the idea of refreshing the metrics right before the prom-handler 👍 
Prometheus will alert when this fails

File: api/pkg/exporters/s3/exporter.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+		if err := http.ListenAndServe(fmt.Sprintf(":%v", listenAddress), nil); err != nil {
+			glog.Fatalf("Failed to listen: %v", err)
+		}
+	}()
+
+	return nil
+}
+
+func (e *s3Exporter) refreshMetrics(w http.ResponseWriter, r *http.Request) bool {
+	prefix := r.URL.Query().Get("prefix")
```
Comment (mrIncompetent) [inline api/pkg/exporters/s3/exporter.go]: So we need to have a dedicated scrape config for each cluster?

File: api/pkg/exporters/s3/exporter.go [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+			e.QuerySuccess.Set(float64(1))
+			return true
+		}
+		objects = append(objects, listerObject)
+	}
+
+	labels := prometheus.Labels{"prefix": prefix}
+	e.ObjectsCount.With(labels).Set(float64(len(objects)))
+	e.ObjectsLastModifiedDate.With(labels).Set(float64(getLastModifiedTimestamp(objects).UnixNano()))
+	e.ObjectsEmptyCount.With(labels).Set(float64(getEmptyObjectCount(objects)))
```
Comment (mrIncompetent) [inline api/pkg/exporters/s3/exporter.go]: Would it be easier if we include the filesize as label on `ObjectsCount` ?

**Comment by metalmatze** [comment]
Before going any further into the code. How often do we want to scrape the S3 bucket?
It shouldn't be that expensive, right?

So in that case I highly recommend using the collector pattern for this exporter. It has the advantage of all the metrics being stateless, which means that for every time we scrape this exporter it will fetch all the data it needs, return the response and then forget about it again.
This is contrary to what the current approach does which has the prometheus client go state.

Using the collector pattern will improve the situation for objects being deleted. Because it's stateless those old objects won't show up anymore whereas with the current approach they would be still in the internal prometheus client_go cache.

Check the docs:
https://godoc.org/github.com/prometheus/client_golang/prometheus#example-Collector
As an example I can point you to my DigitalOcean exporter: https://github.com/metalmatze/digitalocean_exporter/

File: api/cmd/s3-exporter/main.go (hunk 2 above)
Comment (alvaroaleman) [reply api/cmd/s3-exporter/main.go:51]: Block termination until a SIG is send to the process

File: api/pkg/exporters/s3/exporter.go (hunk 8 above)
Comment (alvaroaleman) [reply api/pkg/exporters/s3/exporter.go]: Yes, I actually thought some time about this but I couldn't come up with a better way. Only alternative would be to create an exporter per cluster, but thats IMHO worse

File: api/pkg/exporters/s3/exporter.go (hunk 9 above)
Comment (alvaroaleman) [reply api/pkg/exporters/s3/exporter.go]: The filesize of what? The biggest file? the smallest? The average? The std derivation from the avg? :)

**Comment by alvaroaleman** [comment]
> So in that case I highly recommend using the collector pattern for this exporter. It has the advantage of all the metrics being stateless, which means that for every time we scrape this exporter it will fetch all the data it needs, return the response and then forget about it again.

Makes sense. The exporter already scrapes synchronously but sounds like there are better approaches to implement that behavior than the current one.

**Comment by alvaroaleman** [comment]
@metalmatze So the only difference I see is that in the example you pointed me at, every metric is a `Collector` which collects metrics via `Collect()` - However that means there is one api call per metric, so basically we would here have three apicalls for the exact same data.

Do I understand that correctly? If yes, I would prefer to keep it the way it is.

**Comment by metalmatze** [comment]
In the example I just separated the metrics by domain and I need to do separate calls there anyway. I think with this exporter you can easily do all within one collector and then do one API call. Or do I miss something?

**Comment by alvaroaleman** [comment]
So after talking with @mrIncompetent we decided that if we write an s3 exporter anyways, we can as well put some more business logic into it.

To not require a scrape config per cluster, I will change the exporter to fetch all available clusters and then return metrics for all clusters and not take the `prefix` argument anymore

**Comment by alvaroaleman** [comment]
@mrIncompetent @metalmatze I think all your comments were addressed, PTAL

File: api/cmd/s3-exporter/README.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+A simple exporter for S3-compatible buckets that will export metrics partitioned by Kubermatic cluster names.
+
+It assumes all objects belonging to a given cluster have a prefix of `${CLUSTERNAME}-`.
+
+Avaiable metrics:
+
+```
+# HELP s3_empty_object_count The amount of object with a size of zero
+# TYPE s3_empty_object_count gauge
+s3_empty_object_count{prefix="q8z2hgk94f"} 1
```
Comment (mrIncompetent) [inline api/cmd/s3-exporter/README.md]: Do we need to stick with `prefix`? Why not using `cluster`?

**Comment by alvaroaleman** [comment]
@mrIncompetent readme still contained outdated list of available metrics, I fixed that

File: api/cmd/s3-exporter/main.go (hunk 3 above)
Comment (mrIncompetent) [reply api/cmd/s3-exporter/main.go]: Not fixed yet

File: api/cmd/s3-exporter/README.md (hunk 5 above)
Comment (alvaroaleman) [reply api/cmd/s3-exporter/README.md:37]: Done

**Review by mrIncompetent — APPROVED** [review]

**Review by kron4eg — APPROVED** [review]

**Pull request merged**